src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TaskMethodInvoker.cs

```diff
@@ -74,16 +74,32 @@ public object InvokeEnd(object instance, out object[] outputs, IAsyncResult resu
                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentException(SR.SFxInvalidCallbackIAsyncResult));
             }
 
-            var tuple = invokeTask.Result;
-
-            var task = tuple.Item1 as Task;
-            if (task.IsFaulted)
+            AggregateException ae = null;
+            Tuple<object, object[]> tuple = null;
+            Task task = null;
+            
+            if (invokeTask.IsFaulted)
+            {
+                Fx.Assert(invokeTask.Exception != null, "Task.IsFaulted guarantees non-null exception.");
+                ae = invokeTask.Exception;
+            }
+            else
             {
-                Fx.Assert(task.Exception != null, "Task.IsFaulted guarantees non-null exception.");
+                Fx.Assert(invokeTask.IsCompleted, "Task.Result is expected to be completed");
+                tuple = invokeTask.Result;
+                task = tuple.Item1 as Task;
+                if (task.IsFaulted)
+                {
+                    Fx.Assert(task.Exception != null, "Task.IsFaulted guarantees non-null exception.");
+                    ae = task.Exception;
+                }
+            }
 
+            if (ae != null)
+            {
                 // If FaultException is thrown, we will get 'callFaulted' behavior below.
                 // Any other exception will retain 'callFailed' behavior.
-                throw FxTrace.Exception.AsError<FaultException>(task.Exception);
+                throw FxTrace.Exception.AsError<FaultException>(ae);
```

Comment:
Does this change the exception type that's surfaced outwards? If yes, would anyone key off the exception type to do any processing?